src/net/tcp.rs

```diff
@@ -65,22 +89,40 @@ pub struct TcpClient {
 }
 
 impl TcpClient {
-    /// Creates and returns a TcpClient wrapper from a raw TcpStream. This is so that we can create separate BufReader and BufWriters around the stream.
+    /// Creates and returns a new TcpClient. It makes a few references to the raw stream and wraps them in BufReader and BufWriter for convenience.
     pub fn new(stream: TcpStream) -> TcpClient {
-        // TODO: Handle this better
-        // Note that this does not create a seperate stream for each clone, so any setting changes made on one propagates to the others.
-        let reader = stream.try_clone().unwrap();
-        let writer = stream.try_clone().unwrap();
-        TcpClient {
-            reader: BufReader::new(reader),
-            writer: BufWriter::new(writer),
+        let reader = BufReader::new(stream.try_clone().unwrap());
+        let writer = BufWriter::new(stream.try_clone().unwrap());
+        let (tx, rx) = channel();
+        TcpClient{
+            reader,
+            writer,
             raw_stream: stream,
-            tx: None,
-            rx: None,
+            tx: Some(tx),
+            rx: Some(rx),
+        }
+    }
+
+    /// Sets up the background loop that waits for data to be received on the rx channel that is meant to be sent to the remote client, then enters a loop to watch for input *from* the remote endpoint.
+    pub fn run(client: Arc<Mutex<TcpClient>>) {
+        if let Ok(mut l) = client.lock() {
```

Comment:
We should return a `Result` here that way we can ensure the integrity of the lock is fine.